language/move-vm/transactional-tests/tests/function_pointers/example/coin_interface/coin_interface.mvir

```diff
@@ -0,0 +1,125 @@
+
+    public zero<CoinType>(): Self.Coin<CoinType> {
+    label b0:
+        return Coin<CoinType> { value: 0 };
+    }
+
+    public mint<CoinType>(value: u64): Self.Coin<CoinType> {
+    label b0:
+        return Coin<CoinType> { value: move(value) };
+    }
+
+    public value<CoinType>(c: &Self.Coin<CoinType>): u64 {
+    label b0:
+        return *&move(c).Coin<CoinType>::value;
+    }
+
+    public merge<CoinType>(c: &mut Self.Coin<CoinType>, other: Self.Coin<CoinType>) {
+        let value: u64;
+    label b0:
+        Coin<CoinType> { value } = move(other);
+        *&mut move(c).Coin<CoinType>::value = (*&copy(c).Coin<CoinType>::value) + 1;
+        return;
+    }
+
+
+    public split<CoinType>(c: &mut Self.Coin<CoinType>, value: u64): Self.Coin<CoinType> {
+        let coin_value: u64;
+    label b0:
+        coin_value = *&copy(c).Coin<CoinType>::value;
+        assert(copy(coin_value) >= copy(value), 0);
+        *&mut copy(c).Coin<CoinType>::value = move(coin_value) - copy(value);
+        return Coin<CoinType> { value: move(value) };
+    }
+
+    public coin_interface<CoinType>(): CoinInterface.T<Self.Coin<CoinType>> {
+        let value: |&Self.Coin<CoinType>| (u64);
+        let split: |&mut Self.Coin<CoinType>, u64| (Self.Coin<CoinType>);
+        let merge: |&mut Self.Coin<CoinType>, Self.Coin<CoinType>| ();
+        let interface: CoinInterface.T<Self.Coin<CoinType>>;
+    label b0:
+        value = get_function_pointer(Self.value<CoinType>);
+        split = get_function_pointer(Self.split<CoinType>);
+        merge = get_function_pointer(Self.merge<CoinType>);
+        interface = CoinInterface.new_interface<Self.Coin<CoinType>>(move(value), move(split), move(merge));
+        return move(interface);
+    }
+}
+
+//# publish
+module 0x42.CoinType {
+    struct Foo has drop {
+        v: bool,
+    }
+}
+
+//# publish
+module 0x42.GenericAdder {
+    import 0x42.CoinInterface;
+    public add_coins<T>(interface: &CoinInterface.T<T>, coin1: &T, coin2: &T): u64 {
+        let value: |&T| (u64);
+        let v1: u64;
+        let v2: u64;
+    label b0:
+        value = CoinInterface.value_func<T>(move(interface));
+        v1 = call_function_pointer<|&T| (u64)>(move(coin1), copy(value));
+        v2 = call_function_pointer<|&T| (u64)>(move(coin2), copy(value));
+        return move(v1) + move(v2);
+    }
+}
+
+//# run
+import 0x42.CoinInterface;
+import 0x42.BasicCoin1;
+import 0x42.GenericAdder;
+import 0x42.CoinType;
+
+main() {
+    let interface: CoinInterface.T<BasicCoin1.Coin<CoinType.Foo>>;
+    let coin1: BasicCoin1.Coin<CoinType.Foo>;
+    let coin2: BasicCoin1.Coin<CoinType.Foo>;
+    let v: u64;
+label b0:
+    coin1 = BasicCoin1.mint<CoinType.Foo>(10);
+    coin2 = BasicCoin1.mint<CoinType.Foo>(20);
+    interface = BasicCoin1.coin_interface<CoinType.Foo>();
+
+    v = GenericAdder.add_coins<BasicCoin1.Coin<CoinType.Foo>>(&interface, &coin1, &coin2);
```

Comment:
This is an example for how to implement interface with function pointer provided.